### DISCUSSION TITLE
How to correctly use slider-crank transmission for cylinders?

### DISCUSSION THREAD

**Discussion by rohit-kumar-j** [google-deepmind/mujoco#374] [Asking for Help] [answered]
I have a `equality/connect` for linking a part my robot,(ref image) the rod of the cylinder is connected to the arm via an equality constaraint like so: (similar to a [crank-slider](https://user-images.githubusercontent.com/37873142/177708178-5c481011-7a16-4a0e-87b5-5b84f00a5ee8.png))

<details>
    <Summary>XML Structure</Summary>

```xml
<body name="parent_1">
    <geom name="parent_1_geom" type="mesh" mesh="parent_1_mesh" />
    <body name="arm" >
        <joint name="joint_arm" type="hinge" />
        <geom name="arm_geom" type="mesh" mesh="arm_mesh" />
    </body>
    <body name="parent_2" >
        <joint name="connecting_joint" type="hinge" />
        <geom name="parent_2_geom" type="mesh" mesh="parent_2_mesh"/>
        <body name="rod" >
            <joint name="joint_rod" type="slide"/>
            <geom name="rod_geom" type="mesh" mesh="rod_mesh" />
        </body>
    </body>
</body>

<equality>
    <connect name="connect_1" body1="rod" body2="arm" anchor=" some_x some_y some_z" />
</equality>

<actuator>
    <position name="position_actuator_rod" joint="joint_rod" kp="100" />
    <position name="position_actuator_arm" joint="joint_arm" kp="100" />
</actuator>
```

</details>

### 1.Parent_1, 2.Parent_2, 3. Rod, 4. Arm
<img src="https://user-images.githubusercontent.com/37873142/177708464-e0d68fe2-582c-45be-9b41-0850a942b389.png" height="400" />

I use the following to control the rod to move the arm, the equality seems to disappear:
<details>
    <Summary>Equality/connect seems to disappear</Summary>

```python
id = mujoco.mj_name2id(
        model,
        mujoco.mjtObj.mjOBJ_ACTUATOR,
        "position_actuator_rod") # This joint is slide, inbetween body 2 and 3
data.ctrl[id] = some_changing_value
```

https://user-images.githubusercontent.com/37873142/177709491-137ee6d8-540c-428c-ba1a-a56db5f05711.mp4

</details>

<details>
    <Summary>Equality/connect works the other way</Summary>

```python
id = mujoco.mj_name2id(
        model,
        mujoco.mjtObj.mjOBJ_ACTUATOR,
        "position_actuator_arm") # This joint is revolute, inbetween body 4 and 1
data.ctrl[id] = some_changing_value
```

https://user-images.githubusercontent.com/37873142/177709288-0e7573e8-6a76-470c-bc2d-40befcf87450.mp4

</details>

**Comment by yuvaltassa** [comment]
I'll look more carefully soon at the details, but if you are trying to implement a slider-crank you should know that MuJuCo has a native slider-crank transmission type (search the docs, I'm currently AFK), equality constraints not required.

**Comment by yuvaltassa** [comment]
Having taken a look, I can't really reproduce what you are seeing so can't really diagnose. If you are still experiencing a strange issue, please attach a full loadable model as a zip file and I'll take a look.

However as I said, if you have slider-cranks in your model you should use the build-in slider-crank transmission mechanism, as it adds no DoFs or equality constraints to your model and is significantly more efficient than your current approach (which is also valid, just less efficient). Here is a minimal model implementing a slider-crank. Note that it has exactly 1 DoF. Load it in `simulate` and play with it to understand how it works.

```xml
<mujoco>
  <worldbody>
    <site name="slidersite" pos="0 -.1 0" zaxis="0 1 0" size=".012" rgba="1 0 0 1"/>
    <body>
      <joint damping=".1"/>
      <geom type="capsule" size=".01" fromto="0 0 0 .2 0 0"/>
      <site name="cranksite" pos=".1 0 0" size=".012" rgba="1 0 0 1"/>
    </body>
  </worldbody>

  <actuator>
    <position name="slider-crank" cranksite="cranksite" slidersite="slidersite" cranklength=".11" 
      ctrllimited="true" ctrlrange="-.1 .1" kp="30"/>
  </actuator>
</mujoco>
```

slider-crank transmissions are currently not well documented. Let us know if you end up using this and we might add better documentation.

**Reply by rohit-kumar-j** [reply]
A labeled diagram with the slider-crank assembly with the terms `cranklength` and slider position would be helpful.

**Comment by rohit-kumar-j** [comment]
I apologize for the inaccurate explanation on my part,
I think the slider-crank in mujoco is similar to [this](https://technologystudent.com/cams/crkslid1.htm):
<image src="https://user-images.githubusercontent.com/37873142/177934916-a5f9c913-0c53-4781-8768-f9506db61d9f.png" width="600">

I wanted to create this:
<image src="https://user-images.githubusercontent.com/37873142/177935346-09fe8e55-eaed-410f-ac59-0afe2bb41e16.png" width="600">
Do you have an example xml for the same?

**Comment by yuvaltassa** [comment, marked as answer]
MuJoCo's mechanism supports both of these. 
Did you actually load my example model in `simulate` and play with it?

Here is a model that shows how to make both forward and backward slider-cranks, and includes an example of a "broken" mechanism which cannot maintain the requested relationship at the initial state.

<details>
  <summary>educational slider-crank model</summary>

```XML
<mujoco model="slider-crank">
  <!-- these are the visual properties associated with the slider-crank mechanism: -->
  <visual>
    <scale slidercrank=".1"/>
    <rgba slidercrank="0.5 0.4 0.8 1" crankbroken="0.0 0.6 0.2 1"/>
  </visual>

  <!-- defining some defaults for readability -->
  <default>
    <position ctrllimited="true" ctrlrange="-.1 .1" kp="30"/>
    <default class="joint">
      <geom type="cylinder" size=".006" fromto="0 0 0 0 0 .05" rgba=".9 .6 1 1"/>
    </default>
    <default class="slidersite">
      <site type="cylinder" size=".01 .015" rgba="1 0 0 1" />
    </default>
    <default class="cranksite">
      <site size=".012" rgba="0 0 1 1" />
    </default>
  </default>

  <worldbody>
    <light pos="0 0 1"/>
    <geom type="plane" size=".5 .5 .01" pos="0 0 -.05"/>
    <site name="slidersite" class="slidersite" pos="0 -.1 0" zaxis="1 .5 0"/>
    <body>
      <joint damping=".1"/>
      <geom class="joint"/>
      <geom type="capsule" size=".01" fromto="0 0 0 .2 0 0"/>
      <site name="cranksite" class="cranksite" pos=".1 0 0"/>
      <site name="cranksite2" class="cranksite" pos=".2 0 0"/>
      <body pos=".15 0 0">
        <joint damping=".1"/>
        <geom class="joint"/>
        <geom type="capsule" size=".01" fromto="0 -.15 0 0 0 0"/>
        <site name="slidersite2" class="slidersite" pos=".02 -.1 0" euler="-90 30 0"/>
      </body>
    </body>
    <site name="slidersite3" class="slidersite" pos="0 .1 0" zaxis="1 .2 0"/>
    <body pos="0 .2 0">
      <joint  damping=".1"/>
      <geom class="joint"/>
      <geom type="capsule" size=".01" fromto="0 0 0 .15 0 0"/>
      <site name="cranksite3" class="cranksite" pos=".1 0 0" size=".012"/>
    </body>
  </worldbody>

  <actuator>
    <!-- forward slider-crank: the slider site is attached to the parent -->
    <position name="forward" cranksite="cranksite" slidersite="slidersite" cranklength=".08"/>
    <!-- backward slider-crank: the slider site is attached to the child -->
    <position name="backward" cranksite="cranksite2" slidersite="slidersite2" cranklength=".06"/>
    <!--  The "broken-crank" actuator is "broken" in the initial state: the crank length is too
          short to maintain the requested geometric relationship -->
    <position name="broken" cranksite="cranksite3" slidersite="slidersite3" cranklength=".05"/>
  </actuator>
</mujoco>
```

</details>

Here is a riddle to check your understanding. The orientation of `slidersite` is defined using `zaxis`, but the orientation of `slidersite2` is defined using `euler`. Do you understand why I had to do that? What happens if you try to use `zaxis` to orient `slidersite2`? In order to answer this question you'll need to use `simulate`'s site-frame visualisation.

**Reply by yuvaltassa** [reply]
Having played with it a bit more I realised the same configuration can be created using the `zaxis` attribute, `euler` is not actually required, but the important lesson remains -- you need to think about the precise orientation of the `slidersite`'s frame.

**Reply by yuvaltassa** [reply]
Note, I added [this model](https://github.com/deepmind/mujoco/blob/main/model/slider_crank/slider_crank.xml)  to the repo.

**Comment by rohit-kumar-j** [comment]
Thank you for the second example(broken-crank), it cleared much confusion.
I did play around with the slider-crank, but with the very first model I ended up with this:

<details>
    <Summary>Confused from trail from 1st example</Summary>

```xml
<mujoco>
    <worldbody>
        <body name="fixed_arm" pos="0 -.1 0">
            <geom type="box" size=".01 .09 .01" rgba="1 0 0 .5" />
            <body name="pink_capsule" pos="0 -.1 0" euler="90 153.5 0">
                <joint axis="0 1 0" />
                <geom type="capsule" size=".02 .03" pos="0 0 .1" rgba="1 0 1 .5" />
                <site name="slidersite" pos="0 0 0" zaxis="0 0 1" size=".012" rgba="1 0 0 1" />
                <body name="rod" pos="0 0 .15">
                    <joint name="slide_joint" type="slide" axis="0 0 1" />
                    <geom type="capsule" size=".015 .025" rgba="1 1 .1 1" />
                </body>
            </body>
            <body pos="0 .1 0">
                <joint name="main_pivot" damping=".1" range="-0.5 1.6" limited="true" />
                <geom type="capsule" size=".01" fromto=".02 0 0 .2 0 0" />
                <site name="cranksite" pos=".1 0 0" size=".012" rgba="1 0 0 1" />
                <body name="connecting_box" pos=".1 0 0">
                    <joint />
                    <geom type="box" size=".01 .01 .01" />
                </body>
            </body>
        </body>
    </worldbody>

    <actuator>
        <position name="slider-crank" cranksite="cranksite" slidersite="slidersite" cranklength=".11" kp="200" />
        <!-- ctrllimited="true" ctrlrange="-1 1" kp="30" -->
        <motor name="force_act" joint="slide_joint" gear="1" ctrllimited="true" ctrlrange="-100 100" />
    </actuator>
    <equality>
        <weld body1="connecting_box" body2="rod" />
    </equality>
    <contact>
        <exclude body1="fixed_arm" body2="pink_capsule" />
    </contact>
</mujoco>
```

https://user-images.githubusercontent.com/37873142/178030756-284a82cc-045c-4333-b466-188b3a3bbea9.mp4

</details>

<details>
    <Summary>But ended up having a Misalignment issue when the links were rotated(prehaps due to soft constraints?):</Summary>

https://user-images.githubusercontent.com/37873142/178030922-6d4baa7c-60f5-4503-bd96-96ee56be2402.mp4

</details>

Albeit, I did not fully understand the way the slider-crank is supposed to be used. Which part of the generated geom was the slider, and which part was the crank? Why does one of them switch from red to purple as I adjust the actuator value? Why/how is the force from this silder crank not working? etc

The 2nd example of the slider-crank, i.e the _educational slider-crank model_ and your question related to the `euler` vs `zaxis` cleared up  some of this confusion.

From [frame-orientations/zaxis](https://mujoco.readthedocs.io/en/latest/modeling.html#frame-orientations):
> This determines the X and Y axes of the frame implicitly. This is useful for geoms with rotational symmetry around the Z axis, as well as lights - which are oriented along the Z axis of their frame.

This makes it easier to use `zaxis` instead of `euler` as the orientation is simpler to adjust. The broken slider-crank example (especially with the orientation of the capsule geom) helped me understand why the cylindrical part of the `silder-crank` behaves a certain way and how to orient the slider-crank.

| Broken Crank | Site Frame | Upon Changing `zaxis` to 1 .5 0 (the solution is feasible and the crank is no longer broken) |
| :-------: | :------: | :------: |
![](https://user-images.githubusercontent.com/37873142/178043151-022dae1e-d836-4e68-bc38-5f67868979e8.png)  |  ![](https://user-images.githubusercontent.com/37873142/178043541-4db5cfcd-210d-4a38-811a-378792370505.png) | ![](https://user-images.githubusercontent.com/37873142/178043872-4b8707a8-bd89-4810-bc63-b6bb31576dbd.png)

______________________

## Solving for Crank

| Feasible Solution(the crank is purple, yellow mark is the point of solution) | Feasible Solution Not Feasible (the crank is green, no solution for the visually generated crank, the capsule body rotates, followed by crank color change ,followed by slider shortening when the simulation is played) |
| :-------: | :------: |
| ![](https://user-images.githubusercontent.com/37873142/178046923-3c9a3aac-fe78-459f-b902-ba639f3ea1cf.png) | ![](https://user-images.githubusercontent.com/37873142/178047435-d4195cf3-2b00-420a-858a-7423941d0f8e.png)

If I understand this correctly, mujoco solves and moves the geom for all possible radii of the cranks(black circles) to intersect to the Zaxis of of the slider(purple cylindrical part of the slider-crank) until the `cranklength` specified in the xml file is reached, at which point, the crankarm turns purple again. Then mujoco prioritizes to reduce the slider position to zero while keeping the `cranklength` as specified in the xml.

The 1st few seconds of the video below, it appears as if the joint between the green(unsolved) crank and the capsule geom does not change, only after the geom itself is rotated so as not to violate the specified `cranklength` and the crank color has turned purple, 

1. The slider moves (length of the cylinder reduces to set point, which is zero)
2. The angle between the crank and the geom varies as a consequence of the slider going to it set point(zero)

### Video:

<details>
    <Summary> Broken crank: @ timestep=1e-6 </Summary>

https://user-images.githubusercontent.com/37873142/178048502-0734fa69-9bf1-4ff1-b49c-45851f7eb56f.mp4

</details>

So, Ideally for a robot, 

1. Should we make it such that the slider-crank look something like [this](https://user-images.githubusercontent.com/37873142/178049785-04b9f68b-4349-4764-a282-644a0a6c84fd.png) at startup? (Assuming, we do not want the sudden jerk to pull the slider crank towards each other)
2. If the site geom does not have the `inertial` properties, how can we add mass/collision properties from the real robot while not violating soft constraints as seen in [Issue: Equality constraint Bounciness](https://github.com/deepmind/mujoco/issues/278) or the misalignment video above?

<img src="https://user-images.githubusercontent.com/37873142/178049785-04b9f68b-4349-4764-a282-644a0a6c84fd.png" width="500" />

**Reply by rohit-kumar-j** [reply]
Renaming from: 
"Equality/connect + data.ctrl don't work together" to "How to correctly use slider-crank transmission for cylinders?"

**Comment by rohit-kumar-j** [comment]
Can we lock joint at the `slidersite`?

![image](https://user-images.githubusercontent.com/37873142/178058499-fdd675e2-fc87-427c-bc70-b4aa51e712f1.png)

**Reply by yuvaltassa** [reply]
Just delete the joint? Or add `<joint limited="true" range="min max"/>`, if that's what you're trying to achieve?

**Reply by yuvaltassa** [reply]
Oh, I see what you're asking. That black circle not at the `slidersite`, that's the connection point of the slider-crank mechanism. 

No. You cannot put a limit on that, you can only put joint ranges on the actual joints.

**Comment by yuvaltassa** [comment]
If it helps, here is the slider-crank [visualisation code](https://github.com/deepmind/mujoco/blob/main/src/engine/engine_vis_visualize.c#L1360) and here is the code [computing the transmission](https://github.com/deepmind/mujoco/blob/main/src/engine/engine_core_smooth.c#L704) (length + moment arm).

Try to think of it in the opposite direction that the force is going. MuJoCo is trying to compute the function: `slider_length(joint_angles)`. Given some joint angles and a predefined crank length `cranklength` (the "crank" is the _thin purple cylinder_ that is connected to the `cranksite`), and given the direction of the `slidersite` z-axis, MuJoCo tries to compute the length of the slider (the _thick purple cylinder_ that is connected to the `slidersite`)) – in closed-form. 
1. If closed-form is not possible, the crank is "broken" and is coloured in a different colour (I made it green in my example, the default is [red](https://mujoco.readthedocs.io/en/latest/XMLreference.html#visual-rgba)). 
2. If it is possible, this length is the `length` of the actuator. This length can be controlled by e.g. a `<position>` actuator.

(The reason the function needs to be closed-form is that we care about its Jacobian (the "moment-arm matrix"), not just its value)

**Comment by rohit-kumar-j** [comment]
For future readers, playing with the time-step, `solref` and `solimp`, a combination of a lower time step and a `equality/weld` did the trick for me:

<details>
    <Summary>Video</Summary>

https://user-images.githubusercontent.com/37873142/180267465-9a918a44-7f48-42f6-bce3-16d29d6ecea1.mp4

</details>


<details>
    <Summary>XML</Summary>

```xml
<mujoco model="slider-crank">
    <option timestep="0.0001" gravity="0 0 -9.81" collision="predefined" />

    <worldbody>
        <light pos="0 0 1" />
        <geom type="plane" size="3 3 .01" pos="0 0 -.2" />
        <body name="b_root" pos="0 0 2" euler="0 90 0">
            <geom name="g_root_1" type="box" fromto="0 0 0 1 0 0" size="0.05" rgba="0 1 .1 .5" mass="0.0001" />
            <geom name="g_root_2" type="cylinder" fromto="1 0 -.1 1 0 .1" size="0.08" rgba="0 1 .1 1" mass="0.0001" />
            <body name="b_arm" pos="1 0 0">
                <joint name="hinge_joint_j1" type="hinge" />
                <geom name="g_arm_1" type="cylinder" fromto="0 0 -.1 0 0 .1" size="0.08" rgba="1 0 .1 1" mass="0.0001" />
                <geom name="g_arm_2" type="box" fromto="0 0 0 0 1 0" size="0.05" rgba="1 0 .1 .5" mass="0.0001" />
                <geom name="g_arm_3" pos="0 1 0" type="sphere" size="0.03" rgba="1 0 .1 1" mass="1" />
            </body>
            <body name="b_cyl" pos=".2 0 0" euler="0 0 -45">
                <joint name="j2" type="hinge" />
                <geom name="g_cyl_1" type="cylinder" fromto="0 0 -.1 0 0 .1" size="0.03" rgba="1 0 .1 1" mass="0.0001" />
                <geom name="g_cyl_2" type="cylinder" fromto="0 0 0 0 0.565685424949238 0" size="0.05" rgba="1 0 .1 .5" mass="0.0001" />
                <body name="b_piston" pos="0 0.565685424949238 0">
                    <joint name="slide_joint_j3" type="slide" axis="0 1 0" />
                    <geom name="g_piston_1" type="cylinder" fromto="0 0.565685424949238 -.1 0 0.565685424949238 .1" size="0.03" rgba="1 0 .1 1" mass="0.0001" />
                    <geom name="g_piston_2" type="cylinder" fromto="0 0 0 0 0.565685424949238 0" size="0.02" rgba="1 0 .1 .5" mass="0.0001" />
                    <body name="weld_me" pos="0 0.565685424949238 0" euler="0 0 45">
                        <joint name="weld_j" />
                        <geom name="g_piston_weld_2" type="cylinder" fromto="0 0 -.1 0 0 .1" size="0.03" rgba="1 1 .1 1" mass="0.0001" />
                    </body>
                </body>
            </body>
        </body>
        <body name="b_root_1" pos="1 0 2" euler="0 90 0">
            <geom name="g_root_1_1" type="box" fromto="0 0 0 1 0 0" size="0.05" rgba="0 .1 1 .5" mass="0.0001" />
            <geom name="g_root_2_2" type="cylinder" fromto="1 0 -.1 1 0 .1" size="0.08" rgba="0 1 .1 1" mass="0.0001" />
            <body name="b_arm_1" pos="1 0 0">
                <joint name="hinge_joint_j1_1" type="hinge" />
                <geom name="g_arm_1_1" type="cylinder" fromto="0 0 -.1 0 0 .1" size="0.08" rgba="1 0 .1 1" mass="0.0001" />
                <geom name="g_arm_2_1" type="box" fromto="0 0 0 0 1 0" size="0.05" rgba="1 0 .1 .5" mass="0.0001" />
                <geom name="g_arm_3_1" pos="0 1 0" type="sphere" size="0.03" rgba="1 0 .1 1" mass="1" />
            </body>
            <body name="b_cyl_1" pos=".2 0 0" euler="0 0 -45">
                <joint name="j2_1" type="hinge" />
                <geom name="g_cyl_1_1" type="cylinder" fromto="0 0 -.1 0 0 .1" size="0.03" rgba="1 0 .1 1" mass="0.0001" />
                <geom name="g_cyl_2_1" type="cylinder" fromto="0 0 0 0 0.565685424949238 0" size="0.05" rgba="1 0 .1 .5" mass="0.0001" />
                <body name="b_piston_1" pos="0 0.565685424949238 0">
                    <joint name="slide_joint_j3_1" type="slide" axis="0 1 0" />
                    <geom name="g_piston_1_1" type="cylinder" fromto="0 0.565685424949238 -.1 0 0.565685424949238 .1" size="0.03" rgba="1 0 .1 1" mass="0.0001" />
                    <geom name="g_piston_2_1" type="cylinder" fromto="0 0 0 0 0.565685424949238 0" size="0.02" rgba="1 0 .1 .5" mass="0.0001" />
                    <body name="weld_me_1" pos="0 0.565685424949238 0" euler="0 0 45">
                        <joint name="weld_j_1" />
                        <geom name="g_piston_weld_2_1" type="cylinder" fromto="0 0 -.1 0 0 .1" size="0.03" rgba="1 1 .1 1" mass="0.0001" />
                    </body>
                </body>
            </body>
        </body>
    </worldbody>
    <equality>
        <weld name="weld_1" active="true" body1="b_arm" body2="weld_me" relpose="0 0.8 0 1 0 0 0" solref="0.0002 1" />
        <connect name="connect_1_1" active="true" body1="weld_me_1" body2="b_arm_1" anchor="0 0.565685424949238 0" />

    </equality>
    <actuator>
        <!-- <position name="j1_act" joint="hinge_joint_j1" kp="1000" /> -->
        <!-- <position name="j1_act_1" joint="hinge_joint_j1_1" kp="1000" /> -->
        <position name="j3_act" joint="slide_joint_j3" kp="1000" />
        <position name="j3_act_1" joint="slide_joint_j3_1" kp="10" />
    </actuator>
    <sensor>
        <jointpos name="j1_sensor" joint="hinge_joint_j1" />
        <jointpos name="j1_sensor_1" joint="hinge_joint_j1_1" />
    </sensor>
</mujoco>
```

</details>
